framework/Optimizers/GradientDescent.py

```diff
@@ -0,0 +1,840 @@
+    self._stepHistory[traj] = deque(maxlen=self._maxHistLen)
+    self._acceptHistory[traj] = deque(maxlen=self._maxHistLen)
+    self._stepRecommendations[traj] = None
+    self._acceptRerun[traj] = False
+    self._convergenceInfo[traj] = {'persistence': 0}
+    for criteria in self._convergenceCriteria:
+      self._convergenceInfo[traj][criteria] = False
+    return traj
+
+  def _resolveNewGradPoint(self, traj, rlz, optVal, info):
+    """
+      Consider and store a new gradient evaluation point
+      @ In, traj, int, trajectory for this new point
+      @ In, info, dict, identifying information about the realization
+      @ In, rlz, dict, realized realization
+      @ In, optVal, float, value of objective variable (corrected for min/max)
+    """
+    self._stepTracker[traj]['grads'].append((rlz, info))
+
+  # * * * * * * * * * * * * * * * *
+  # Resolving potential opt points
+  def _applyBoundaryConstraints(self, point):
+    """
+      Checks and fixes boundary constraints of variables in "point" -> DENORMED point expected!
+      @ In, point, dict, potential point against which to check
+      @ Out, point, dict, adjusted variables
+      @ Out, modded, bool, whether point was modified or not
+    """
+    # TODO should some of this go into the parent Optimizer class, such as the boundary acquiring?
+    modded = False
+    for var in self.toBeSampled:
+      dist = self.distDict[var]
+      val = point[var]
+      lower = dist.lowerBound
+      upper = dist.upperBound
+      if lower is None:
+        lower = -np.inf
+      if upper is None:
+        upper = np.inf
+      if val < lower:
+        self.raiseADebug(' BOUNDARY VIOLATION "{}" suggested value: {:1.3e} lower bound: {:1.3e} under by {:1.3e}'
+                         .format(var, val, lower, lower - val))
+        self.raiseADebug(' ... -> for point {}'.format(point))
+        point[var] = lower
+        modded = True
+      elif val > upper:
+        self.raiseADebug(' BOUNDARY VIOLATION "{}" suggested value: {:1.3e} upper bound: {:1.3e} over by {:1.3e}'
+                         .format(var, val, upper, val - upper))
+        self.raiseADebug(' ... -> for point {}'.format(point))
+        point[var] = upper
+        modded = True
+    return point, modded
+
+  def _applyFunctionalConstraints(self, suggested, previous):
+    """
+      @ In, suggested, dict, NORMALIZED suggested point
+      @ In, previous, dict, NORMALIZED previous point
+      @ Out, suggested, dict, fixed up normalized point
+      @ Out, modded, bool, True if point was modified within this method
+    """
+    # assume no modifications until proved otherwise
+    modded = False
+    # are we violating functional constraints?
+    passFuncs = self._checkFunctionalConstraints(self.denormalizeData(suggested))
+    # while in violation of constraints ...
+    info = {'minStepSize': self._convergenceCriteria.get('stepSize', 1e-10)} # TODO why 1e-10?
+    tries = 500
```

Comment:
This would be an easy feature to add later instead of delaying this PR. I suspect it will be rare for a user to want to modify this, and it doesn't involve model runs, so I consider this a low-priority user option.

Reply:
Unfortunately we are used to "forget" these minimal things and they stay in the code forever. Even if it is detail, I think we should had this parameter in the InputData container for the GradientDescent now and recreate, automatically, the manual portion.

Reply:
I think it would come up the first time someone wanted to control it, and until then it's time spent on low-value low-impact improvements. I will add it.